File: app/src/main/java/com/iceteaviet/fastfoodfinder/ui/profile/ProfileFragment.java
package com.iceteaviet.fastfoodfinder.ui.profile;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.iceteaviet.fastfoodfinder.App;
import com.iceteaviet.fastfoodfinder.R;
import com.iceteaviet.fastfoodfinder.data.DataManager;
import com.iceteaviet.fastfoodfinder.data.remote.user.model.User;
import com.iceteaviet.fastfoodfinder.data.remote.user.model.UserStoreList;
import com.iceteaviet.fastfoodfinder.utils.Constant;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;
import butterknife.BindView;
import butterknife.ButterKnife;
import de.hdodenhof.circleimageview.CircleImageView;
import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;

import static com.iceteaviet.fastfoodfinder.ui.profile.ListDetailActivity.KEY_USER_PHOTO_URL;
import static com.iceteaviet.fastfoodfinder.ui.profile.ListDetailActivity.KEY_USER_STORE_LIST;


public class ProfileFragment extends Fragment {
    @BindView(R.id.ivCoverImage)
    ImageView ivCoverImage;
    @BindView(R.id.btnUpdateCoverImage)
    Button btnUpdateCoverImage;
    @BindView(R.id.ivCreate)
    CircleImageView civCreate;
    @BindView(R.id.cvCreateNew)
    CardView cvCreate;
    @BindView(R.id.iv_profile_avatar)
    ImageView ivAvatarProfile;
    @BindView(R.id.tvName)
    TextView tvName;
    @BindView(R.id.tvEmail)
    TextView tvEmail;
    @BindView(R.id.rvListPacket)
    RecyclerView rvListPacket;
    @BindView(R.id.tvNumberList)
    TextView tvNumberList;
    @BindView(R.id.cv_saved_places)
    CardView cvSavePlace;
    @BindView(R.id.cv_checkin_places)
    CardView cvCheckinPlace;
    @BindView(R.id.cv_favourite_places)
    CardView cvFavouritePlace;
    @BindView(R.id.fav_list_items_count)
    TextView tvFavItemsCount;

    private DialogUpdateCoverImage mDialog;
    private DialogCreateNewList mDialogCreate;
    private UserStoreListAdapter mAdapter;
    private List<UserStoreList> defaultList;
    private ArrayList<String> listName;

    private DataManager dataManager;

    public static ProfileFragment newInstance() {
        Bundle extras = new Bundle();
        ProfileFragment fragment = new ProfileFragment();
        fragment.setArguments(extras);
        return fragment;
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);

        listName = new ArrayList<>();
        defaultList = new ArrayList<>();
        dataManager = App.getDataManager();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_profile, container, false);
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mAdapter = new UserStoreListAdapter();
        StaggeredGridLayoutManager mLayoutManager = new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL);
        rvListPacket.setAdapter(mAdapter);
        rvListPacket.setLayoutManager(mLayoutManager);

        tvName.setText(R.string.unregistered_user);
        tvEmail.setText(R.string.unregistered_email);
        if (dataManager.getCurrentUser() == null)
            dataManager.setCurrentUser(new User(getString(R.string.unregistered_user), getString(R.string.unregistered_email), Constant.NO_AVATAR_PLACEHOLDER_URL, "null", new ArrayList<UserStoreList>()));

        if (dataManager.isSignedIn()) {
            getCurrentUserData();
        }
    }

    public void loadUserList() {
        User currentUser = dataManager.getCurrentUser();
        for (int i = 0; i < currentUser.getUserStoreLists().size(); i++) {
            if (i <= 2) {
                defaultList.add(currentUser.getUserStoreLists().get(i));
            } else {
                mAdapter.addListPacket(currentUser.getUserStoreLists().get(i));
            }
        }
        tvNumberList.setText(String.format("(%s)", String.valueOf(mAdapter.getItemCount())));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        MenuItem item = menu.findItem(R.id.action_search);
        item.setVisible(false);
    }

    private void getCurrentUserData() {
        dataManager.getRemoteUserDataSource().getUser(dataManager.getCurrentUserUid())
                .subscribe(new SingleObserver<User>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onSuccess(User user) {
                        dataManager.setCurrentUser(user);
                        Glide.with(getContext())
                                .load(user.getPhotoUrl())
                                .into(ivAvatarProfile);
                        tvName.setText(user.getName());
                        tvEmail.setText(user.getEmail());
                        loadUserList();
                        for (int i = 0; i < user.getUserStoreLists().size(); i++) {
                            listName.add(user.getUserStoreLists().get(i).getListName());
                        }
                        tvFavItemsCount.setText(String.format(Locale.getDefault(), "%d nơi", user.getFavouriteStoreList().getStoreIdList().size()));
                        onListener();
                    }

                    @Override
                    public void onError(Throwable e) {
                        e.printStackTrace();
                    }
                });
    }

    public void onListener() {
        ivCoverImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                btnUpdateCoverImage.setVisibility(View.VISIBLE);
            }
        });

        btnUpdateCoverImage.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mDialog = DialogUpdateCoverImage.newInstance();
                mDialog.setOnButtonClickListener(new DialogUpdateCoverImage.OnButtonClickListener() {
                    @Override
                    public void onButtonClickListener(int Id, Bitmap bitmap) {
                        if (Id != 0)
                            if (Id == -1) {
                                ivCoverImage.setImageBitmap(bitmap);
                            } else {
                                ivCoverImage.setImageResource(Id);
                            }
                    }
                });
                mDialog.show(getFragmentManager(), "");

                btnUpdateCoverImage.setVisibility(View.GONE);
            }
        });

        civCreate.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View view) {
                mDialogCreate = DialogCreateNewList.newInstance(listName);
                mDialogCreate.show(getFragmentManager(), "");
                mDialogCreate.setOnButtonClickListener(new DialogCreateNewList.OnCreateListListener() {
                    @Override
                    public void onButtonClick(String name, int idIconSource) {
                        User currentUser = dataManager.getCurrentUser();
                        int id = currentUser.getUserStoreLists().size(); //New id = current size
                        UserStoreList list = new UserStoreList(id, new ArrayList<Integer>(), idIconSource, name);
                        mAdapter.addListPacket(list);
                        currentUser.addStoreList(list);
                        dataManager.getRemoteUserDataSource().updateStoreListForUser(currentUser.getUid(), currentUser.getUserStoreLists());
                        tvNumberList.setText(String.format("(%s)", String.valueOf(mAdapter.getItemCount())));
                    }
                });
            }
        });

        cvSavePlace.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sendToDetailListActivity(defaultList.get(UserStoreList.ID_SAVED));

            }
        });
        cvFavouritePlace.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sendToDetailListActivity(defaultList.get(UserStoreList.ID_FAVOURITE));

            }
        });
        cvCheckinPlace.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                sendToDetailListActivity(defaultList.get(UserStoreList.ID_CHECKED_IN));
            }
        });

        mAdapter.setOnItemLongClickListener(new UserStoreListAdapter.OnItemLongClickListener() {
            @Override
            public void onClick(int position) {
                User currentUser = dataManager.getCurrentUser();
                tvNumberList.setText(String.format("(%s)", String.valueOf(mAdapter.getItemCount())));
                currentUser.removeStoreList(position);
                dataManager.getRemoteUserDataSource().updateStoreListForUser(currentUser.getUid(), currentUser.getUserStoreLists());
            }
        });

        mAdapter.setOnItemClickListener(new UserStoreListAdapter.OnItemClickListener() {
            @Override
            public void onClick(UserStoreList listPacket) {
                sendToDetailListActivity(listPacket);

            }
        });
    }


    private void sendToDetailListActivity(UserStoreList userStoreList) {
        Intent intent = new Intent(getContext(), ListDetailActivity.class);
        intent.putExtra(KEY_USER_PHOTO_URL, dataManager.getCurrentUser().getPhotoUrl());
        intent.putExtra(KEY_USER_STORE_LIST, userStoreList);
        startActivity(intent);
    }
}

File: app/src/main/kotlin/com/iceteaviet/fastfoodfinder/ui/profile/ProfileFragment.kt
package com.iceteaviet.fastfoodfinder.ui.profile

import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.view.*
import android.widget.ImageView
import androidx.annotation.NonNull
import androidx.annotation.Nullable
import androidx.cardview.widget.CardView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.bumptech.glide.Glide
import com.iceteaviet.fastfoodfinder.App
import com.iceteaviet.fastfoodfinder.R
import com.iceteaviet.fastfoodfinder.data.DataManager
import com.iceteaviet.fastfoodfinder.data.remote.user.model.User
import com.iceteaviet.fastfoodfinder.data.remote.user.model.UserStoreList
import com.iceteaviet.fastfoodfinder.ui.custom.store.StoreListView
import com.iceteaviet.fastfoodfinder.ui.profile.ListDetailActivity.Companion.KEY_USER_PHOTO_URL
import com.iceteaviet.fastfoodfinder.ui.profile.ListDetailActivity.Companion.KEY_USER_STORE_LIST
import com.iceteaviet.fastfoodfinder.utils.isValidUserUid
import com.iceteaviet.fastfoodfinder.utils.openLoginActivity
import io.reactivex.SingleObserver
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.fragment_profile.*
import java.util.*


class ProfileFragment : Fragment(), View.OnClickListener {
    lateinit var ivAvatarProfile: ImageView
    lateinit var cvSavePlace: StoreListView
    lateinit var cvFavouritePlace: StoreListView
    lateinit var btnCreateNew: CardView

    private var mDialog: DialogUpdateCoverImage? = null
    private var mDialogCreate: DialogCreateNewList? = null
    private var mAdapter: UserStoreListAdapter? = null
    private var defaultList: MutableList<UserStoreList> = ArrayList() // Default store list (saved, favourite) that every user have
    private var listName: ArrayList<String> = ArrayList()

    private lateinit var dataManager: DataManager


    override fun onCreate(@Nullable savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)

        dataManager = App.getDataManager()
    }

    @Nullable
    override fun onCreateView(@NonNull inflater: LayoutInflater, @Nullable container: ViewGroup?, @Nullable savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_profile, container, false)
    }

    override fun onViewCreated(@NonNull view: View, @Nullable savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        setupUi()
        setupEventListeners()
        loadCurrentUserData()
    }

    override fun onResume() {
        super.onResume()
        // Invalid auth token -> go to login screen
        if (!dataManager.isSignedIn())
            openLoginActivity(activity!!)
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.cvCreateNew -> {
                mDialogCreate = DialogCreateNewList.newInstance(listName)
                mDialogCreate!!.show(fragmentManager!!, "")
                mDialogCreate!!.setOnButtonClickListener(object : DialogCreateNewList.OnCreateListListener {
                    override fun onButtonClick(name: String, iconId: Int) {
                        val currentUser = dataManager.getCurrentUser()
                        val id = currentUser!!.getUserStoreLists().size //New id = current size
                        val list = UserStoreList(id, ArrayList(), iconId, name)
                        mAdapter!!.addListPacket(list)
                        currentUser.addStoreList(list)
                        dataManager.getRemoteUserDataSource().updateStoreListForUser(currentUser.uid, currentUser.getUserStoreLists())
                        tvNumberList!!.text = String.format("(%s)", mAdapter!!.itemCount.toString())
                    }
                })
                return
            }

            R.id.cv_saved_places -> {
                sendToDetailListActivity(defaultList[UserStoreList.ID_SAVED])
                return
            }

            R.id.cv_favourite_places -> {
                sendToDetailListActivity(defaultList[UserStoreList.ID_FAVOURITE])
                return
            }

            R.id.btnUpdateCoverImage -> {
                mDialog!!.show(fragmentManager!!, "")
                btnUpdateCoverImage!!.visibility = View.GONE
                return
            }
        }
    }

    private fun setupEventListeners() {
        btnCreateNew.setOnClickListener(this)
        cvSavePlace.setOnClickListener(this)
        cvFavouritePlace.setOnClickListener(this)
        btnUpdateCoverImage.setOnClickListener(this)

        mAdapter!!.setOnItemLongClickListener(object : UserStoreListAdapter.OnItemLongClickListener {
            override fun onClick(position: Int) {
                val currentUser = dataManager.getCurrentUser()
                tvNumberList!!.text = String.format("(%s)", mAdapter!!.itemCount.toString())
                currentUser!!.removeStoreList(position)
                dataManager.getRemoteUserDataSource().updateStoreListForUser(currentUser.uid, currentUser.getUserStoreLists())
            }
        })

        mAdapter!!.setOnItemClickListener(object : UserStoreListAdapter.OnItemClickListener {
            override fun onClick(listPacket: UserStoreList) {
                sendToDetailListActivity(listPacket)
            }
        })

        mDialog!!.setOnButtonClickListener(object : DialogUpdateCoverImage.OnButtonClickListener {
            override fun onOkClick(Id: Int, bmp: Bitmap?) {
                if (Id != 0)
                    if (Id == -1) {
                        ivCoverImage!!.setImageBitmap(bmp)
                    } else {
                        ivCoverImage!!.setImageResource(Id)
                    }

                btnUpdateCoverImage!!.visibility = View.VISIBLE
            }

            override fun onCancelClick() {
                btnUpdateCoverImage!!.visibility = View.VISIBLE
            }
        })
    }

    private fun setupUi() {
        ivAvatarProfile = iv_profile_avatar
        cvSavePlace = cv_saved_places
        cvFavouritePlace = cv_favourite_places
        btnCreateNew = cvCreateNew

        mAdapter = UserStoreListAdapter()
        val mLayoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        rvListPacket!!.adapter = mAdapter
        rvListPacket!!.layoutManager = mLayoutManager

        mDialog = DialogUpdateCoverImage.newInstance()

        tvName!!.setText(R.string.unregistered_user)
        tvEmail!!.setText(R.string.unregistered_email)
    }

    fun loadStoreLists() {
        val currentUser = dataManager.getCurrentUser()
        for (i in 0 until currentUser!!.getUserStoreLists().size) {
            if (i <= 2) {
                defaultList.add(currentUser.getUserStoreLists()[i])
            } else {
                mAdapter!!.addListPacket(currentUser.getUserStoreLists()[i])
            }
        }
        tvNumberList!!.text = String.format("(%s)", mAdapter!!.itemCount.toString())
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        val item = menu.findItem(R.id.action_search)
        item.isVisible = false
    }

    private fun loadCurrentUserData() {
        val uid = dataManager.getCurrentUserUid()
        if (!isValidUserUid(uid))
            return

        dataManager.getRemoteUserDataSource().getUser(uid)
                .subscribe(object : SingleObserver<User> {
                    override fun onSubscribe(d: Disposable) {

                    }

                    override fun onSuccess(user: User) {
                        dataManager.setCurrentUser(user)
                        Glide.with(context!!)
                                .load(user.photoUrl)
                                .into(ivAvatarProfile)
                        tvName!!.text = user.name
                        tvEmail!!.text = user.email
                        loadStoreLists()
                        for (i in 0 until user.getUserStoreLists().size) {
                            listName.add(user.getUserStoreLists()[i].listName)
                        }

                        fillStoreListPreviewData(user)
                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                    }
                })
    }

    private fun fillStoreListPreviewData(user: User) {
        cvSavePlace.setCount(user.getSavedStoreList().getStoreIdList()!!.size.toString())
        cvFavouritePlace.setCount(user.getFavouriteStoreList().getStoreIdList()!!.size.toString())
    }

    fun sendToDetailListActivity(userStoreList: UserStoreList) {
        val intent = Intent(context, ListDetailActivity::class.java)
        intent.putExtra(KEY_USER_PHOTO_URL, dataManager.getCurrentUser()!!.photoUrl)
        intent.putExtra(KEY_USER_STORE_LIST, userStoreList)
        startActivity(intent)
    }

    companion object {
        fun newInstance(): ProfileFragment {
            val extras = Bundle()
            val fragment = ProfileFragment()
            fragment.arguments = extras
            return fragment
        }
    }
}
